src/lib/snapp.ts

```diff
@@ -0,0 +1,327 @@
+      });
+    },
+
+    assertEquals(a: A) {
+      if (!this._initialized)
+        throw Error(
+          'assertEquals can only be called when the State is assigned to a SmartContract @state.'
+        );
+      const r = this.getLayout();
+      const xs = this._ty.toFields(a);
+      let e: ExecutionState = this._this.executionState();
+
+      xs.forEach((x, i) => {
+        e.party.predicate.state[r.offset + i].check = new Bool(true);
+        e.party.predicate.state[r.offset + i].value = x;
+      });
+    },
+
+    async get(): Promise<A> {
+      if (!this._initialized)
+        throw Error(
+          'get can only be called when the State is assigned to a SmartContract @state.'
+        );
+      const r = this.getLayout();
+
+      let addr: PublicKey = this._this.address;
+      let p: Promise<Field[]>;
+
+      if (Circuit.inProver()) {
+        p = Mina.getAccount(addr).then((a) => {
+          const xs: Field[] = [];
+          for (let i = 0; i < r.length; ++i) {
+            xs.push(a.snapp.appState[r.offset + i]);
+          }
+          return Circuit.witness(Circuit.array(Field, r.length), () => xs);
+        });
+      } else {
+        const res = Circuit.witness(Circuit.array(Field, r.length), () => {
+          throw 'unimplemented';
+        });
+        p = new Promise((k) => k(res));
+      }
+
+      let xs = await p;
+      const res = this._ty.ofFields(xs);
+      if ((this._ty as any).check != undefined) {
+        (this._ty as any).check(res);
+      }
+      return res;
+    },
+  };
+}
+
+type InternalStateType = ReturnType<typeof createState>;
+
+/**
+ * A decorator to use within a snapp to indicate what will be stored on-chain.
+ * For example, if you want to store a field element `some_state` in a snapp,
+ * you can use the following in the declaration of your snapp:
+ *
+ * ```
+ * @state(Field) some_state = State<Field>();
+ * ```
+ *
+ */
+export function state<A>(ty: AsFieldElements<A>) {
+  return function (
+    target: any,
+    key: string,
+    _descriptor?: PropertyDescriptor
+  ): any {
+    const SnappClass = target.constructor;
+    if (!(SnappClass.prototype instanceof SmartContract)) {
+      throw new Error(
+        'Can only use @state decorator on classes that extend SmartContract'
+      );
+    }
+
+    // TBD: ok to not check? bc type metadata not inferred from class field assignment
+    // const fieldType = Reflect.getMetadata('design:type', target, key);
+    // if (fieldType != State) {
+    //   throw new Error(
+    //     `@state fields must have type State<A> for some type A, got ${fieldType}`
+    //   );
+    // }
+
+    if (key === '_states' || key === '_layout') {
+      throw new Error('Property names _states and _layout reserved.');
+    }
+
+    if (SnappClass._states == undefined) {
+      SnappClass._states = [];
+      let layout: Map<string, { offset: number; length: number }>;
+      SnappClass._layout = () => {
```

Comment:
I wonder why we are setting `_states` and `_layout` dynamically rather than keeping it simple and having these as class fields on SmartContract. We could even have private class fields with the `#` syntax!

Reply:
Aaah nvm. Of course it has to be dynamic since it's a property of the specific class that extends SmartContract